### ISSUE TITLE
Reflowing text selection styles when editing

### ISSUE THREAD

**Issue by GreenFootballs** [fabricjs/fabric.js#6133]
In an app I'm developing, we allow the user to create graphics by entering text in an HTML textarea, then rendering that text in a fabricjs Textbox. We then allow the user to select text in the textarea and apply styles to the selection, then render it again and use setSelectionStyles() with the start and end indexes of the selected area. It works great, but if the user then edits the text in the textarea, inserts, deletes, etc., the added styles don't reflow with the changes -- they remain at the indexes, even though the text has moved around. So it tends to look pretty strange.

I noticed in [this demo](http://fabricjs.com/kitchensink) that the same thing happens if you edit the text in the textarea, but **not** if you edit the text directly on the fabric canvas. On the canvas it reflows automatically.

So my question: is there any way to get the same reflowing behavior for selection styles, while editing in a textarea?

**Comment by asturur** [comment]
i m not sure it can be done. Did you try looking at the events that fabricjs attach to the hidden textarea, if they can be attached to your custom textarea, if they do any good?

**Comment by GreenFootballs** [comment]
I'll have a look at them. Can you tell me where this is in the code?

**Comment by asturur** [comment]
look for 'onInput:' and then trace where that event is attached or removed. Look for the lifecycle of hiddenTextarea and see if it can leave outside of the canvas.

**Comment by GreenFootballs** [comment]
Interesting. Found [the onInput handler](https://github.com/fabricjs/fabric.js/blob/c954b0bea5cdd9c69576e9df011c32507e9890c6/src/mixins/itext_key_behavior.mixin.js#L137) and this does look like it's adjusting the positions of selection styles according to changes. I'll have to look deeper to see if any of it can be extracted to use with an external textarea.

One question: is 'this' in the code referring to the fabric canvas instance?

**Comment by GreenFootballs** [comment]
This hiddenTextarea object is where the actual editing happens when you're in edit mode, correct? One idea might be to trigger edit mode somehow, then find the hiddenTextarea, copy the value of the external textarea to it, and trigger a keyup event to get it to execute the onInput handler.

Is there a way to programmatically trigger edit mode on a Textbox?

**Comment by asturur** [comment]
would be better to just understand how to handle it also externally rather than buildind a fragile solution.

**Comment by GreenFootballs** [comment]
Yes, I think I agree. Thanks for the pointer to the onInput code - it will take a bit of work but that's what I needed.

**Comment by asturur** [comment]
please share your ideas/findings/experiments/test results.
I'll try to help to find the edge cases you may not think of for lack of internal knowledge of the library.

We could end up with a mergeable pr for everyone

**Comment by GreenFootballs** [comment]
I'll be sure to let you know if I have anything to show.

It would be great if one of the options for creating an IText or Textbox was an HTML textarea to bind to that object for editing purposes. If the options isn't set, create the hiddenTextarea  as it does now.

**Comment by asturur** [comment]
Hi @GreenFootballs did you do any discovery/progress on this?

**Comment by GreenFootballs** [comment]
Sorry, I got involved in other areas of the project and haven't had time yet. I do plan to look into it eventually, though.

**Comment by GreenFootballs** [comment]
I'm getting back into this code now, and I have a couple of questions about the onInput handler:

1. What is the "fabric" variable referencing at line 194? The global fabric object doesn't seem to have the properties it uses.

2. I think I need to look at the updateFromTextArea() function too, but which file is this in?

**Comment by GreenFootballs** [comment]
Ah, never mind. [Found it.](https://github.com/fabricjs/fabric.js/blob/26e1a5b55cbeeffb59845337ced3f3f91d533d7d/src/mixins/itext_behavior.mixin.js#L477)

Is this the code that renders an editable textarea inside the fabric canvas? Is it necessary to do the stuff in this function at the end of the onInput function, or could I just render the TextBox object after adjusting the style positions?

**Comment by asturur** [comment]
that code `updateFromTextArea` grabs the content of the text area and update the fabricJS text object information to render it.
The textarea is the source of truth for displaying the text, but is displayed by the data in the fabricObject.

**Comment by GreenFootballs** [comment]
Another question: I notice in [a couple of places](https://github.com/fabricjs/fabric.js/blob/c954b0bea5cdd9c69576e9df011c32507e9890c6/src/mixins/itext_key_behavior.mixin.js#L168), the code references "this.selectionStart" and "this.selectionEnd". Since "this" is the TextBox object, how are these values being updated? In my test code, when I trace to this point, those values are always zero, so they must be updated somewhere else.

Since it's using those values to determine whether text has been added or deleted, my test code isn't quite working.

**Comment by asturur** [comment]
selectionStart and selectionEnd with the `this` keyword are updated in 3 cases.
From mouse operations on the canvas, and mirrored into the textarea, then from the textarea read again and rendered.

Or updated from arrow keys in fabricJS and then mirrored in the textarea, read again from textarea and rendered, also the shortcut for select ALL is handled by fabricJS i believe

All the other kind of updates are just read from the textarea.

**Comment by GreenFootballs** [comment]
I've got it working! A few issues still to resolve though - some cases where it still fails to properly re-flow the styles.

This testing code is part of a Chrome extension under development, so it isn't something I can easily share, but I'll set up a Codepen example when I have time.

I still think it would be a good idea to pass an optional textarea element in the constructor for the TextBox object, and have it bind to that instead of creating a hidden textarea. From looking at the code it seems like this wouldn't require too many changes.

**Comment by asturur** [comment]
no it would not require much changes, i m surprised style does not reflow automatically.
